plugins/lib/plugin.ts

```diff
@@ -0,0 +1,164 @@
+import net from "node:net";
+import { resolve } from "node:path";
+import { v4 as uuidv4 } from "uuid";
+
+type SendTransactionPayload = {
+  to: string;
+  value: string;
+  data: string;
+  gas_limit: number;
+  gas_price?: number;
+  speed?: string;
+  max_fee_per_gas?: number;
+  max_priority_fee_per_gas?: number;
+  valid_until?: number;
+};
+
+type SendTransactionResult = {
+  tx_hash: string;
+};
+
+type Result<T> = {
+  request_id: string;
+  result: T;
+  error: string | null;
+};
+
+type Relayer = {
+  sendTransaction: (
+    payload: SendTransactionPayload
+  ) => Promise<Result<SendTransactionResult>>;
+};
+
+function getPluginParams(): unknown {
+  const pluginParams = process.argv[3];
+
+  if (pluginParams) {
+    try {
+      return JSON.parse(pluginParams);
+    } catch (e) {
+      throw new Error(`Failed to parse payload: ${e}`);
+    }
+  } else return {};
+}
+
+export async function runPlugin(
+  main: (plugin: Plugin, pluginParams: unknown) => Promise<void>
+) {
+  try {
+    // checks if socket path is provided
+    let socketPath = process.argv[2];
+    if (!socketPath) {
+      throw new Error("Socket path is required");
+    }
+
+    // creates plugin instance
+    let plugin = new Plugin(socketPath);
+
+    // runs main function
+    await main(plugin, getPluginParams())
+      .then(() => plugin.close())
+      .catch((error) => {
+        console.error(error);
+        // closes socket signaling error
+        plugin.closeErrored(error);
+      })
+      .finally(() => {
+        plugin.close();
+        process.exit(0);
+      });
+  } catch (error) {
+    console.error(error);
+    process.exit(1);
+  }
+}
+
+export class Plugin {
+  socket: net.Socket;
+  pending: Map<
+    string,
+    { resolve: (value: any) => void; reject: (reason: any) => void }
+  >;
+  private _connectionPromise: Promise<void> | null = null;
+  private _connected: boolean = false;
+
+  constructor(socketPath: string) {
+    this.socket = net.createConnection(socketPath);
+    this.pending = new Map();
+
+    this._connectionPromise = new Promise((resolve, reject) => {
+      this.socket.on("connect", () => {
+        this._connected = true;
+        resolve();
+      });
+
+      this.socket.on("error", (error) => {
+        console.error("Socket ERROR:", error);
+        reject(error);
+      });
+    });
+
+    this.socket.on("data", (data) => {
+      data
+        .toString()
+        .split("\n")
+        .filter(Boolean)
+        .forEach((msg: string) => {
+          const parsed = JSON.parse(msg);
+          const { requestId, result, error } = parsed;
+          const resolver = this.pending.get(requestId);
+          if (resolver) {
+            error ? resolver.reject(error) : resolver.resolve(result);
+            this.pending.delete(requestId);
+          }
+        });
+    });
+  }
+
+  useRelayer(relayerId: string): Relayer {
```

Comment:
do we have plan to introduce other methods to plugin system, like get relayer, list relayers...?